pandas/_libs/groupby.pyx

```diff
@@ -700,6 +700,117 @@ group_mean_float32 = _group_mean['float']
 group_mean_float64 = _group_mean['double']
 
 
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _group_mean_transposed(floating[:, :] out,
+                           int64_t[:] counts,
+                           ndarray[floating, ndim=2] values,
+                           const int64_t[:] labels,
+                           Py_ssize_t min_count=-1):
+    cdef:
+        Py_ssize_t i, j, N, K, lab, ncounts = len(counts)
+        floating val, count, y, t
+        floating[:, :] sumx, compensation
+        int64_t[:, :] nobs
+        Py_ssize_t len_values = len(values), len_labels = len(labels)
+
+    assert min_count == -1, "'min_count' only used in add and prod"
+
+    # if len_values != len_labels:
+    #     raise ValueError("len(index) != len(labels)")
+
+    nobs = np.zeros((<object>out).shape, dtype=np.int64)
+    sumx = np.zeros_like(out)
+    compensation = np.zeros_like(out)
+
+    K, N = (<object>values).shape
+
+    if N != len_labels:
+        raise ValueError("len(index) != len(labels)")
+
+    with nogil:
+        for i in range(N):
+            lab = labels[i]
+            if lab < 0:
+                continue
+
+            counts[lab] += 1
+            for j in range(K):
+                val = values[j, i]
```

Comment:
would it make a difference if we defined `temp_values = values[:, i]` (or values[i] in non-transposed version) outside the j-loop and then set `val = temp_values[j]` inside the loop?

definitely seems weird that this would have a big perf impact